### ISSUE TITLE
Wrong loss or lambda works wrong

### ISSUE THREAD

**Issue by stas-semenov** [keras-team/keras#12200]
When I try to run this code, I have the wrong loss.

```
import numpy as np
from keras.layers import Input, Dense, Lambda
from keras.models import Model
import keras.backend as K

def train_iterator(batch_size=16):
    x = np.zeros((batch_size, 2))
    while True:
        for i in range(batch_size):
            x[i][0] = np.random.randint(0, 100)
            x[i][1] = np.random.randint(0, 100)
        x_mean = (x[::,0] + x[::,1]) / 2
        yield [x], [x_mean]

K.clear_session()
x = Input(name = 'x', shape=(2,))
x_mean = Lambda(lambda x: (x[::,0] + x[::,1]) / 2, output_shape=(1,), name = 'x_mean')(x)
model = Model(inputs=x, outputs=x_mean)
model.summary()

model.compile(loss=['mse'], optimizer = 'adam')
model.fit_generator(train_iterator(), steps_per_epoch = 10, epochs = 5, verbose = 1)

x, x_mean = next(train_iterator(1))
print(x, x_mean, model.predict(x))
```
Epoch 1/5
10/10 [==============================] - 0s 7ms/step - loss: 829.3230
Epoch 2/5
10/10 [==============================] - 0s 2ms/step - loss: 936.7182
Epoch 3/5
10/10 [==============================] - 0s 2ms/step - loss: 667.0758
Epoch 4/5
10/10 [==============================] - 0s 2ms/step - loss: 757.9758
Epoch 5/5
10/10 [==============================] - 0s 2ms/step - loss: 1024.0672

**Comment by stas-semenov** [comment]
Doesn't work:
```
def train_iterator(batch_size=16):
    x = np.zeros((batch_size, 2))
    while True:
        for i in range(batch_size):
            x[i][0] = np.random.randint(0, 100)
            x[i][1] = np.random.randint(0, 100)
        x_res = (x[::,0] + x[::,1]) / 2
        yield [x], [x_res]
```
Works perfect:
```
def train_iterator(batch_size=16):
    x = np.zeros((batch_size, 2))
    x_res = np.zeros((batch_size, 1))
    while True:
        for i in range(batch_size):
            x[i][0] = np.random.randint(0, 100)
            x[i][1] = np.random.randint(0, 100)
        x_res[::,0] = (x[::,0] + x[::,1]) / 2
        yield [x], [x_res]
```
Squeezed version also works fine:
```
def train_iterator(batch_size=16):
    x = np.zeros((batch_size, 2))
    x_res = np.zeros((batch_size, 1))
    while True:
        for i in range(batch_size):
            x[i][0] = np.random.randint(0, 100)
            x[i][1] = np.random.randint(0, 100)
        x_res[::,0] = (x[::,0] + x[::,1]) / 2
        x_res_sq = np.squeeze(x_res, axis=-1)
        yield [x], [x_res_sq]
```
